speedy/model/model.jl

```diff
@@ -0,0 +1,820 @@
+
+    ist::T
+    jst::T
+
+end
+
+@metadata name ("","","","","","") NTuple{6, String}
+@metadata unit ("","","","","","") NTuple{6, String}
+@metadata description ("","","","","","") NTuple{6, String}
+
+@name @description @unit struct StateVectors{T<:AbstractArray, S<:AbstractArray}
+
+    particles::T | ("u","v","T","q","ps","rain") | ("velocity x-component","velocity y-component","Temperature", "Specific Humidity", "Pressure", "Rain") | ("m/s","m/s","K","kg/kg","Pa","mm/hr")
+    truth::S | ("ps") | ("Pressure") | ("kPa")
+
+end
+
+function get_axes(params::ModelParameters)
+
+    return get_axes(params.nlon, params.nlat, params.nlev)
+
+end
+function get_axes(ix::Int, iy::Int, iz::Int)
+    x = LinRange(0, 2*π, ix)
+    y = LinRange(0, π, iy)
+    z = LinRange(0, 1, iz)
+
+    return x,y,z
+end
+
+function init_gaussian_random_field_generator(params::ModelParameters)
+
+    x, y, z = get_axes(params)
+    return init_gaussian_random_field_generator(params.lambda,params.nu, params.sigma, x, y, params.padding, params.primes)
+
+end
+
+# Initialize a gaussian random field generating function using the Matern covariance kernel
+# and circulant embedding generation method
+# TODO: Could generalise this
+function init_gaussian_random_field_generator(lambda::Vector{T},
+                                              nu::Vector{T},
+                                              sigma::Vector{T},
+                                              xpts::AbstractVector{T},
+                                              ypts::AbstractVector{T},
+                                              pad::Int,
+                                              primes::Bool) where T
+
+    # Let's limit ourselves to two-dimensional fields
+    dim = 2
+    function _generate(l, n, s)
+        cov = Periodic_Cov.CovarianceFunction(dim, Periodic_Cov.CustomDistanceCovarianceStructure(Periodic_Cov.SphericalDistance(), Matern(l, n, σ = s)))
+        grf = GaussianRandomField(cov, Spectral(), xpts, ypts)
+        v = grf.data.eigenval
+        xi = Array{eltype(grf.cov)}(undef, size(v)..., nthreads())
+        w = Array{complex(float(eltype(grf.cov)))}(undef, size(v)..., nthreads())
+        z = Array{eltype(grf.cov)}(undef, length.(grf.pts)..., nthreads())
+        RandomField(grf, xi, w, z)
+    end
+
+    return [_generate(l, n, s) for (l, n, s) in zip(lambda, nu, sigma)]
+end
+# Get a random sample from random_field_generator using random number generator rng
+function sample_gaussian_random_field!(field::AbstractArray{T,2},
+                                       random_field_generator::RandomField,
+                                       rng::Random.AbstractRNG) where T
+
+    field .= GaussianRandomFields.sample(random_field_generator.grf)
+    # @. @view(random_field_generator.xi[:,:,threadid()]) = randn((rng,), T)
+    # sample_gaussian_random_field!(field, random_field_generator, @view(random_field_generator.xi[:,:,threadid()]))
```

Comment:
It'd be good to remove all the commented out code :wink:

Reply:
```suggestion
```